Numpy 2.0 incompatibility

To report a non-security related issue, please provide:

* the version of the software with which you are encountering an issue: current `master` branch
* environmental information (i.e. Operating System, compiler info, java version, python version, etc.): Python 3.12 with numpy 2.0 nightly wheel
* a description of the issue with the steps needed to reproduce it:

Numpy 2.0 will now require that you call the C API level `np.import_array()` before using the C API in Cython extensions. Importing cftime with numpy 2.0 gives you:

```
../../../miniconda3/envs/test-environment/lib/python3.12/site-packages/cftime/__init__.py:1: in <module>
    from ._cftime import (datetime, real_datetime,
src/cftime/_cftime.pyx:1: in init cftime._cftime
    ???
E   ImportError: numpy.core.multiarray failed to import (auto-generated because you didn't call 'numpy.import_array()' after cimporting numpy; use '<void>numpy._import_array' to disable if you are certain you don't need it).
```

Note: Numpy 2.0 isn't out yet...but it will be.

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -37,7 +37,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.16.5
       env:
         # Skips pypy py36,37
-        CIBW_SKIP: "pp* cp36-* cp37-* cp38-*"
+        CIBW_SKIP: "pp* cp36-* cp37-*"
         CIBW_ARCHS: ${{ matrix.arch }}
         CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
         CIBW_TEST_SKIP: "*_arm64"

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -46,9 +46,6 @@ jobs:
           python -c "import cftime; print(f'cftime v{cftime.__version__}')" &&
           python -m pip install -r {package}/requirements-dev.txt &&
           python -m pytest -vv {package}/test
-        # below only for building against unstable numpy
-        CIBW_BUILD_FRONTEND: "build; args: --no-isolation"
-        CIBW_BEFORE_BUILD: "pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy cython setuptools versioneer"
 
     - uses: actions/upload-artifact@v3
       with:

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -44,7 +44,7 @@ jobs:
         CIBW_TEST_REQUIRES: pytest
         CIBW_TEST_COMMAND: >
           python -c "import cftime; print(f'cftime v{cftime.__version__}')" &&
-          python -m pip install check-manifest cython pytest &&
+          python -m pip install check-manifest cython pytest pytest-cov &&
           python -m pytest -vv {package}/test
 
     - uses: actions/upload-artifact@v3

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -37,7 +37,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.16.5
       env:
         # Skips pypy py36,37
-        CIBW_SKIP: "pp* cp36-* cp37-*"
+        CIBW_SKIP: "pp* cp36-* cp37-* cp38-*"
         CIBW_ARCHS: ${{ matrix.arch }}
         CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
         CIBW_TEST_SKIP: "*_arm64"

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -37,7 +37,7 @@ jobs:
       uses: pypa/cibuildwheel@v2.15.0
       env:
         # Skips pypy py36,37
-        CIBW_SKIP: "pp* cp36-* cp37-* 
+        CIBW_SKIP: "pp* cp36-* cp37-*"
         CIBW_ARCHS: ${{ matrix.arch }}
         CIBW_BUILD_FRONTEND: build
         CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -39,7 +39,6 @@ jobs:
         # Skips pypy py36,37
         CIBW_SKIP: "pp* cp36-* cp37-*"
         CIBW_ARCHS: ${{ matrix.arch }}
-        CIBW_BUILD_FRONTEND: build
         CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
         CIBW_TEST_SKIP: "*_arm64"
         CIBW_TEST_REQUIRES: pytest
@@ -47,6 +46,9 @@ jobs:
           python -c "import cftime; print(f'cftime v{cftime.__version__}')" &&
           python -m pip install -r {package}/requirements-dev.txt &&
           python -m pytest -vv {package}/test
+        # below only for building against unstable numpy
+        CIBW_BUILD_FRONTEND: "build; args: --no-isolation"
+        CIBW_BEFORE_BUILD: "pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy cython setuptools versioneer"
 
     - uses: actions/upload-artifact@v3
       with:

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -34,7 +34,7 @@ jobs:
         fetch-depth: 0
 
     - name: "Building ${{ matrix.os }} (${{ matrix.arch }}) wheels"
-      uses: pypa/cibuildwheel@v2.15.0
+      uses: pypa/cibuildwheel@v2.16.5
       env:
         # Skips pypy py36,37
         CIBW_SKIP: "pp* cp36-* cp37-*"

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -44,7 +44,7 @@ jobs:
         CIBW_TEST_REQUIRES: pytest
         CIBW_TEST_COMMAND: >
           python -c "import cftime; print(f'cftime v{cftime.__version__}')" &&
-          python -m pip install -r {package}/requirements-dev.txt &&
+          python -m pip install check-manifest cython pytest &&
           python -m pytest -vv {package}/test
 
     - uses: actions/upload-artifact@v3

--- a/.github/workflows/tests_conda.yml
+++ b/.github/workflows/tests_conda.yml
@@ -46,11 +46,13 @@ jobs:
         --trusted-host pypi.anaconda.org \
         --no-deps --pre --upgrade \
         numpy;
+        python -m pip install -v -e  . --no-deps --no-build-isolation --force-reinstall
 
     - name: Install cftime
+      if: matrix.experimental != true
       shell: bash -l {0}
       run: |
-        pip install -v -e  . --no-deps --force-reinstall
+        python -m pip install -v -e  . --no-deps --force-reinstall
 
     - name: Run Tests
       shell: bash -l {0}

--- a/.github/workflows/tests_conda.yml
+++ b/.github/workflows/tests_conda.yml
@@ -14,9 +14,14 @@ jobs:
         python-version: [ "3.8", "3.9", "3.10", "3.11", "3.12" ]
         os: [windows-latest, ubuntu-latest, macos-latest]
         platform: [x64, x32]
+        experimental: [false]
         exclude:
           - os: macos-latest  
             platform: x32
+        include:
+          - python-version: "3.12"
+            os: "ubuntu-latest"
+            experimental: true
 
     steps:
     - uses: actions/checkout@v4
@@ -31,6 +36,16 @@ jobs:
           cython>=0.29.20
           pytest
           pytest-cov
+          -
+    - name: Install unstable dependencies
+      if: matrix.experimental == true
+      shell: bash -l {0}
+      run: |
+        python -m pip install \
+        --index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/ \
+        --trusted-host pypi.anaconda.org \
+        --no-deps --pre --upgrade \
+        numpy;
 
     - name: Install cftime
       shell: bash -l {0}

--- a/.github/workflows/tests_conda.yml
+++ b/.github/workflows/tests_conda.yml
@@ -37,7 +37,6 @@ jobs:
           cython>=0.29.20
           pytest
           pytest-cov
-          -
     - name: Install unstable dependencies
       if: matrix.experimental == true
       shell: bash -l {0}

--- a/.github/workflows/tests_conda.yml
+++ b/.github/workflows/tests_conda.yml
@@ -8,6 +8,7 @@ on:
 jobs:
   run:
     runs-on: ${{ matrix.os }}
+    continue-on-error: ${{ matrix.experimental }}
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/tests_latest.yml
+++ b/.github/workflows/tests_latest.yml
@@ -20,9 +20,15 @@ jobs:
       run: |
         python -m pip install --upgrade pip
 
-    - name: Install cftime dependencies via pip
+    - name: Install unstable cftime dependencies via pip
       run: |
-        python -m pip install -r requirements-dev.txt
+        python -m pip install --pre -r requirements-dev.txt
+        # get nightly wheels for numpy
+        python -m pip install \
+          --index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/ \
+          --trusted-host pypi.anaconda.org \
+          --no-deps --pre --upgrade \
+          numpy
 
     - name: Install cftime
       run: |

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,3 +1,3 @@
 [build-system]
-requires = ["setuptools>=41.2", "cython>=0.29.20", "wheel", "oldest-supported-numpy; python_version<'3.12.0.rc1'", "numpy>=1.26.0b1; python_version>='3.12.0.rc1'"]
+requires = ["setuptools>=41.2", "cython>=0.29.20", "wheel", "numpy>=1.26.0b1"]
 build-backend = "setuptools.build_meta"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,3 +1,9 @@
 [build-system]
-requires = ["setuptools>=41.2", "cython>=0.29.20", "wheel", "numpy>=2.0.0rc1,<3"]
+requires = [
+    "setuptools>=41.2",
+    "cython>=0.29.20",
+    "wheel",
+    "oldest-supported-numpy ; python_version < '3.9'",
+    "numpy>=2.0.0rc1,<3 ; python_version >= '3.9'",
+]
 build-backend = "setuptools.build_meta"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,3 +1,3 @@
 [build-system]
-requires = ["setuptools>=41.2", "cython>=0.29.20", "wheel", "numpy>=1.26.0b1"]
+requires = ["setuptools>=41.2", "cython>=0.29.20", "wheel", "numpy>=2.0.0rc1,<3"]
 build-backend = "setuptools.build_meta"

--- a/src/cftime/_cftime.pyx
+++ b/src/cftime/_cftime.pyx
@@ -7,12 +7,15 @@ from cpython.object cimport (PyObject_RichCompare, Py_LT, Py_LE, Py_EQ,
 from numpy cimport int64_t, int32_t
 import cython
 import numpy as np
+cimport numpy as np
 import re
 import time
 from datetime import datetime as datetime_python
 from datetime import timedelta, MINYEAR, MAXYEAR
 import warnings
 from ._strptime import _strptime
+
+np.import_array()
 
 microsec_units = ['microseconds','microsecond', 'microsec', 'microsecs']
 millisec_units = ['milliseconds', 'millisecond', 'millisec', 'millisecs', 'msec', 'msecs', 'ms']

